### PR TITLE
add pod-delete-rate-limiter

### PR DESCRIPTION
**Pull request by umezawatakeshi**
Signed-off-by: UMEZAWA Takeshi <takeshi-umezawa@cybozu.co.jp>

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1147,6 +1147,9 @@ workflows:
       - build:
           name: build-pause
           container-image: pause
+      - build:
+          name: build-pod-delete-rate-limiter
+          container-image: pod-delete-rate-limiter
```
Comment (masa213f) [inline .circleci/config.yml:1152]: Can we use ghcr?
I want to push the new container to ghcr as much as possible.

**Review by masa213f — CHANGES_REQUESTED** [review]
Please update `maintenance.md` and add `pod-delete-rate-limitter/README.md`.

File: pod-delete-rate-limiter/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# s3gw container
+
+# Stage1: build from source
+FROM quay.io/cybozu/golang:1.20-jammy AS build
+
+COPY . /work
```
[the remark this answers is not in the document]
Comment (umezawatakeshi) [reply pod-delete-rate-limiter/Dockerfile:6]: `/work` is destination.

[the remark this answers is not in the document]
Comment (masa213f) [reply pod-delete-rate-limiter/Dockerfile:6]: You're right!

File: pod-delete-rate-limiter/hooks/rate_limiter_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+func isAllowed(res *admission.Response) bool {
+	return res.Allowed && res.Result.Code == int32(http.StatusOK)
+}
+
+func isDenied(res *admission.Response) bool {
```
[the remark this answers is not in the document]
Comment (umezawatakeshi) [reply pod-delete-rate-limiter/hooks/rate_limiter_test.go:21]: No.
In `admission.Handler` context, the return values are typically created by `Allowd`, `Denied`, and `Errored` functions. `is...` functions in the test code is named after them.

File: maintenance.md [hunk 4]
```diff
@@ -733,6 +733,15 @@ opentelemetry-collector container consists of three repositories: opentelemetry-
 3. Update `K8S_VERSION` and `PAUSE_VERSION` in `Dockerfile`.
 4. Update `BRANCH` and `TAG` files.
 
+## pod-delete-rate-limiter
+
+![Regular Update](./regular_update.svg)
```
Comment (masa213f) [inline maintenance.md:738]: pod-delete-rate-limiter depends on k8s libraries.
So we should update it in the Kubernetes Update.

A version is hard-coded on ENVTEST_K8S_VERSION in a Makefile. we need update it.

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -31,6 +31,8 @@ jobs:
             container-image: "stakater-reloader"
           - dir: "./trust-packages"
             container-image: "trust-packages"
+          - dir: "./pod-delete-rate-limiter"
+            container-image: "pod-delete-rate-limiter"
```
Comment (masa213f) [inline .github/workflows/main.yaml:35]: Will the test is run in CI?

File: pod-delete-rate-limiter/hooks/rate_limiter_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (masa213f) [reply pod-delete-rate-limiter/hooks/rate_limiter_test.go:22]: I understood.

**Review by masa213f — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yaml (hunk 5 above)
Comment (umezawatakeshi) [reply .github/workflows/main.yaml:35]: added a option to run `make test` before building  a image outside of `docker build`.

File: maintenance.md [hunk 6]
```diff
@@ -733,6 +733,27 @@ opentelemetry-collector container consists of three repositories: opentelemetry-
...
+![Kubernetes Update](./kubernetes_update.svg)
+
+1. Update the following version variables in `Makefile`.
+   - `CONTROLLER_TOOLS_VERSION`
+   - `KUSTOMIZE_VERSION`
+   - `ENVTEST_K8S_VERSION`
+2. Update go modules.
+3. Generate code and manifests.
+   ```bash
+   $ cd $GOPATH/src/github.com/cybozu/neco-containers/admission
```
Comment (masa213f) [inline maintenance.md:747]: Please do self-review.
```suggestion
   $ cd $GOPATH/src/github.com/cybozu/neco-containers/pod-delete-rate-limiter
```

**Review by masa213f — CHANGES_REQUESTED** [review]

**Review by masa213f — APPROVED** [review]

**Pull request merged**